bundles/org.openhab.binding.bluetooth.generic/src/main/java/org/openhab/binding/bluetooth/generic/internal/GenericBluetoothHandler.java

```diff
@@ -84,20 +85,51 @@ public void initialize() {
         readCharacteristicJob = scheduler.scheduleWithFixedDelay(() -> {
             if (device.getConnectionState() == ConnectionState.CONNECTED) {
                 if (resolved) {
-                    for (CharacteristicHandler charHandler : charHandlers.values()) {
-                        if (charHandler.canRead()) {
-                            device.readCharacteristic(charHandler.characteristic);
-                            try {
-                                // TODO the ideal solution would be to use locks/conditions and timeouts
-                                // between this code and `onCharacteristicReadComplete` but
-                                // that would overcomplicate the code a bit and I plan
-                                // on implementing a better more generalized solution later
-                                Thread.sleep(50);
-                            } catch (InterruptedException e) {
-                                return;
+                    // go through all characteristics of all services
+                    this.device.getServices().forEach((service) -> {
+                        service.getCharacteristics().forEach((characteristic) -> {
+                            CharacteristicHandler charHandler;
+                            if (device.canNotify(characteristic)) {
+                                // If notifications can be enabled, we create the CharacteristicHandler
+                                // if it not exists so that we can register notifications.
+                                charHandler = charHandlers.computeIfAbsent(characteristic, CharacteristicHandler::new);
+                            } else {
+                                // Otherwise we let it being created by getCharacteristicHandler() and do
+                                // nothing here if it is not created yet.
+                                charHandler = charHandlers.get(characteristic);
                             }
```

Comment:
I don't see why you need to populate the CharacteristicHandler instances here. They should get populated naturally when all of the channels get created, is there a reason that you need to create them sooner?

Reply:
I intend to subscribe for notifications here, when the device is connected. At this point in time the list `charHandlers` does not always contain the `charHandler` already. For this reason I create it with `computeIfAbsent()` so that I can use it some lines below to access data on it.
Is there an other way to access this data or would you put this code in an other place?

Reply:
The notification subscriptions are already getting handled by the `onCharacteristicUpdate` method. You don't need to do anything extra.

Reply:
It is entirely possible that not all of the charHandlers have been created yet. That is to be expected in a multithreaded system. But I don't want multiple parts of the code trying to create it either since there might be addition logic involved and I don't want such logic to be added to multiple places. As long as they system becomes eventually consistent in a short period of time then you don't need to add additional complexity to the code.

Reply:
I didn't see where `onCharacteristicUpdate()` subscribes to notifications. Where is that implemented?

Reply:
`onCharacteristicUpdate()` is part of `BluetoothDeviceListener` which is implemented by the BeaconBluetoothHandler ancestor class. `onCharacteristicUpdate` is called automatically whenever a notification is received.

Reply:
Ah, you talk about the notification part. I was looking for the subscription part, where we activate the subscription on a Characteristic so that it starts notifying.